dali/python/nvidia/dali/fn.py

```diff
@@ -0,0 +1,126 @@
+Got {1} instead when calling operator {2}.""".format(idx, type(inp).__name__, op_class.__name__))
+        default_dev = nvidia.dali.ops._choose_device(args)
+        if default_dev == "gpu" and scalar_args.get("device") == "cpu":
+            raise ValueError("An operator with device='cpu' cannot accept GPU inputs.")
+        if "device" not in scalar_args:
+            scalar_args["device"] = default_dev
+
+        return op_class(**scalar_args)(*args, **tensor_args)
+    return op_wrapper
+
+def _wrap_op(op_class):
+    wrapper_name = _to_snake_case(op_class.__name__)
+    if not hasattr(sys.modules[__name__], wrapper_name):
+        setattr(sys.modules[__name__], wrapper_name, _wrap_op_fn(op_class))
+
+
+def external_source(source = None, num_outputs = None, *, cycle = False, name = None, device = "cpu", layout = None):
+    """
+    Creates a data node which is populated with data from a Python source.
+    The data can be provided by the `source` function or iterable, or it can be provided by
+    `pipeline.feed_input(name, data, layout)` inside `pipeline.iter_setup`.
+
+    `source` : callable or iterable
+    The source of the data. The source is polled for data (via a call or `next(source)` whenever
+    the pipeline needs input for the next iteration. The source can supply one or data batches,
+    depending on the value or `num_outputs`. If `num_outputs` is not set, the `source` is expected
+    to return a single batch. If it's specified, the data is expected to a tuple or list where each
+    element corresponds to respective return value of the external_source.
+
+    `cycle`: bool
+    If `True`, the source iterable will be wrapped. Otherwise, StopIteration error wil be raised
+    when end of data is reached. Setting this flag to True when `source` is not an iterable is an
+    error.
+
+    `num_outputs` : int, optional
+    If specified, denotes the number of TensorLists produced by the source function
+
+    `name` : str, optional
+    The name of the data node - used when feeding the data in `iter_setup`; can be omitted if
+    the data is provided by `source`.
+
+    `layout` : str or list/tuple of str:
+    If provided, sets the layout of the data. When `num_outputs` > 1, layout can be a list
+    containing a distinct layout for each output. If the list has fewer elements than `num_outputs`,
+    only the first outputs have the layout set, the reset have it cleared.
+    """
+    if num_outputs is not None:
+        if source is None:
+            raise ValueError("The parameter `num_outputs` is only valid when using `source` to "
+                "provide data. To feed multiple external sources in `feed_input`, use multiple "
+                "`external_source` nodes.")
+
+    op = nvidia.dali.ops.ExternalSource(device = device, num_outputs = num_outputs,
+                                        source = source, cycle = cycle, layout = layout)
+    return op(name = name)
```

Comment:
Shouldn't the name be passed to the op `__init__` instead of `__call__`?

Reply:
No - it took me by surprise, too, but it makes sense. You can do:
```
op = ops.MyOp(whatever)
x = op(input, name = "x")
y = op(input, name = "y")
```

Reply:
Sure.